[RHSTOR-4798] MDSCacheUsageHigh alert feature test cases automation.

nagendra202 opened the pull request:
[RHSTOR-4798] MDSCacheUsageHigh alert feature test cases automation.

Feature Released: 4.15

Test Plan: https://docs.google.com/spreadsheets/d/172CiopML3LbPC3AN9Dm6lKFzfvbs-Etxj6YTYPBZ9Ew/edit#gid=1005735948



Automated below two tests: [linked to Polarion OCS-5478]
1. [RHSTOR-4798](https://issues.redhat.com//browse/RHSTOR-4798) Verify MDSCacheUsageHigh alert and actionable link.
2. [RHSTOR-4798](https://issues.redhat.com//browse/RHSTOR-4798) Verify alert with node drain.

fbalak commented on kubeconfig:
```diff
@@ -0,0 +1,52 @@
+apiVersion: v1
```
This should not be part of the pr

hnallurv replied:
@nagendra202 please check and address the above comment at the earliest

nagendra202 replied:
Not sure how this file has been pushed. Anyhow deleted it.

hnallurv replied:
@fbalak could you please review the rest and provide your comments?

DanielOsypenko commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py:
```diff
@@ -0,0 +1,102 @@
+
+log = logging.getLogger(__name__)
+
+
```
we miss tier, polarion-id, and squad decorators

nagendra202 replied:
Added tier and squad name just above the class. Polarion id added for all tests.

DanielOsypenko commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,133 @@
+log = logging.getLogger(__name__)
+
+
+@pytest.fixture(scope="function")
```
we miss tier, polarion-id and squad decorators

nagendra202 replied:
Added tier and squad name just above the class. Polarion id added for all tests.

OdedViner commented on ocs_ci/templates/workloads/helper_scripts/meta_data_io.py:
```diff
@@ -0,0 +1,111 @@
+        logging.error(f"Error during metadata operations: {str(e)}")
+
+
+if __name__ == "__main__":
```
Would you like to run this page directly? if yes, why?

nagendra202 replied:
This is a standalone script will run inside the fedora pods for few mins to generate load in cache.

nagendra202 commented on ocs_ci/templates/workloads/helper_scripts/file_creator_io.py:
```diff
@@ -0,0 +1,59 @@
+        return results
+
+
+if __name__ == "__main__":
```
This is a standalone script will run inside the fedora pods for few mins to generate load on CPU.

DanielOsypenko commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,211 @@
+        # Run meta_data_io.py on fedora pod
+        log.info("Running meta data IO on fedora pod ")
+        metaio_executor = ThreadPoolExecutor(max_workers=1)
+        # self.metaio_thread = metaio_executor.submit(
```
comment left

nagendra202 replied:
Removed the commented line of code.

DanielOsypenko commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,211 @@
+        log.info(
+            "Metadata IO started in the background. Lets wait for 15 minutes before validating the MDS alert"
+        )
+        time.sleep(900)
```
Not critical.

do you want to pull these timeouts to module or class variables? 
I see in PR 900, 600, 100 sec. This may help to maintain the module.

nagendra202 replied:
Good suggestion, implemented the same.

DanielOsypenko commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,211 @@
+        self, run_metadata_io_with_cephfs, threading_lock
+    ):
+        # This function verifies the mds cache alert with both active and standby-replay mds scale down and up
+        time.sleep(900)
```
Please add logger.info message before every time.sleep that we are doing, so we can identify timeouts in the log console, including the sleep period. Same as you did at [tests/functional/monitoring/prometheus/alerts/test_alert_mds_cpu_high_usage.py](https://github.com/red-hat-storage/ocs-ci/pull/9757/files#diff-092955c44cfe901fa9242fe90704a92a42b7dcdabb650727da505351ac4a2cea)

 ```
log.info(
            "File creation IO started in the background."
            " Script will sleep for 15 minutes before validating the MDS alert"
        )

```

nagendra202 replied:
Implemented the changes.

PrasadDesala commented on ocs_ci/ocs/cluster.py:
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
+    ct_pod = pod.get_ceph_tools_pod()
+    ceph_mdsmap = ct_pod.exec_ceph_cmd("ceph fs status")
+
+    # Find ceph daemon state as 'active'
```
you can remove the commented code

nagendra202 replied:
done

PrasadDesala commented on ocs_ci/ocs/cluster.py:
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
+
+    logger.info(f"Found active MDS daemon: {ceph_daemon_name}")
+
+    # Find ceph MDS pod running 'active' daemon.
```
all these comments can be removed, instead convert it to log message

nagendra202 replied:
done

PrasadDesala commented on ocs_ci/ocs/cluster.py:
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
+
+    # Find ceph MDS pod running 'active' daemon.
+    mds_pods = get_mds_pods()
+    active_pod = next((srp for srp in mds_pods if ceph_daemon_name in srp.name), None)
```
replace srp with some meaningful name, may be pod?

nagendra202 replied:
done

PrasadDesala commented on ocs_ci/ocs/cluster.py:
```diff
@@ -3316,3 +3316,68 @@ def check_cephcluster_status(
+        )
+        return None
+    # Get the node name of running pod.
+    node_name = active_pod.data["spec"].get("nodeName")
```
use .get("spec") to prevent potential keyerror

PrasadDesala commented on ocs_ci/ocs/constants.py:
```diff
@@ -2862,3 +2865,10 @@
+# helper script to perform meta data operations [create, rename, change permissions...etc] on app pod to fill MDS cache
+METAIO = os.path.join(TEMPLATE_WORKLOAD_DIR, "helper_scripts/meta_data_io.py")
+# helper script to perform file creation IO on app pod to fill MDS cpu
+FILE_CREATOR_IO = os.path.join(
```
Not a blocker for this PR, but IMO we should push this script to an image and run it as a container

PrasadDesala commented on ocs_ci/utility/prometheus.py:
```diff
@@ -783,3 +794,75 @@ def unsubscribe(self):
+        description(str): description of the alert triggered
+        runbook(str): runbook_url in the alert
+        severity(str): severity of the alert
+    """
```
add an empty line above this

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+@tier2
+@blue_squad
+class TestMdsMemoryAlerts:
+    alert_timer = 900  # sleep time to generate the alert 15 minutes
```
add an empty line above this

nagendra202 replied:
added an empty line.

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+    @pytest.mark.polarion_id("OCS-5570")
+    def test_alert_triggered(
+        self, run_metadata_io_with_cephfs, threading_lock, alert_timer
+    ):
```
Please add a description about the test case

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+    def test_mds_cache_alert_with_active_node_drain(
+        self, run_metadata_io_with_cephfs, threading_lock, alert_timer
+    ):
+        # This function verifies the mds cache alert with active mds running node drain
```
Please add a description about the test case

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        assert active_mds_alert_values(threading_lock)
+
+        node_name = cluster.get_active_mds_info()["node_name"]
+        # Unschedule active mds running node.
```
you can remove the comments at all the places since they are repetitive

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+    log.info("Checking for Ceph Health OK")
+    ceph_health_check_base()
+
+    for i in range(3):
```
replace i with some meaningful name

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+    2. Create dc pod with Fedora image
+    3. Copy helper_scripts/meta_data_io.py to Fedora dc pod
+    4. Run meta_data_io.py on fedora pod
+    """
```
add an empty line above

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+    access_mode = constants.ACCESS_MODE_RWX
+    file = constants.METAIO
+    interface = constants.CEPHFILESYSTEM
+    log.info("Checking for Ceph Health OK")
+    ceph_health_check_base()
```
this check is already done at the beginning of the test execution by our framework, I think we can remove it here

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        pvc_obj = pvc_factory(interface=interface, access_mode=access_mode, size="30")
+        # Creating a Fedora dc pod
+        log.info("Creating fedora dc pod")
+        pod_obj = dc_pod_factory(
```
dc_pod_factory can even create a PVC, is there any reason why you are using PVC separately using pvc_factory above?

nagendra202 replied:
done

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        # Copy meta_data_io.py to fedora pod
+        log.info("Copying meta_data_io.py to fedora pod ")
+        cmd = f"oc cp {file} {pvc_obj.namespace}/{pod_obj.name}:/"
+        helpers.run_cmd(cmd=cmd)
```
since you have the pod_obj, you could use copy_to_pod_rsync func to copy the file from a local path to the pod path

nagendra202 replied:
To approach that way, it will be adding few more lines of code and need to import Pod class for that. Here only once we copy the file to the pod. So I feel ok to use the single line command for now. If any issues arise then will look into this?

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        )
+
+
+def active_mds_alert_values(threading_lock):
```
doc string missing

nagendra202 replied:
added doc string

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+def active_mds_alert_values(threading_lock):
+    active_mds = cluster.get_active_mds_info()["mds_daemon"]
+    sr_mds = cluster.get_mds_standby_replay_info()["mds_daemon"]
+    cache_alert = constants.ALERT_MDSCACHEUSAGEHIGH
```
since you are calling it only once, you can directly call the constant in the function

nagendra202 replied:
Anyhow I have to import constants for other variables so I just added it to the alerts section where everyone added the alerts in the constants.py. I can hard code it by just giving alert name within this function. I feel not to do that way because in future if someone want that alert to reuse they can refer alerts section in constant.py

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        "https://github.com/openshift/runbooks/blob/master/alerts/"
+        "openshift-container-storage-operator/CephMdsCacheUsageHigh.md"
+    )
+    state = "firing"
+    severity = "critical"
```
since you are calling it only once, you can directly call the constant in the function

nagendra202 replied:
Anyhow I have to import constants for other variables so I just added it to the alerts section where everyone added the alerts in the constants.py. I can hard code it by just giving alert name within this function. I feel not to do that way because in future if someone want that alert to reuse they can refer alerts section in constant.py

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+        drain_nodes([node_name])
+        log.info(f"node {node_name} drained successfully")
+        # Make the node schedule-able
+        schedule_nodes([node_name])
```
how do you make sure that, the app pod running node is not picked here?

nagendra202 replied:
We need to bring down a node where Active MDS pod is running. So the same  node name given for node drain, the same node name is giving for schedule_nodes as well.

PrasadDesala commented on tests/functional/monitoring/prometheus/alerts/test_alert_mds_cache_high_usage.py:
```diff
@@ -0,0 +1,234 @@
+            deployment_name=deployment_name, replica_count=1
+        )
+        log.info(
+            "Metadata IO started in the background. Script will sleep for 15 minutes before validating the MDS alert"
```
how do you make sure that, the app pod running node is not picked here?

nagendra202 reviewed (commented):
Addressed all review comments and pushed changes.

fbalak reviewed (changes requested):
Other monitoring tests use concept of measurement fixtures: the system is moved into a disturbed state state in fixture, after that there is started alert collecting in a thread to monitor how the system behaves and then the measurement is saved. This way you can use those measurements in multiple test cases and you have logged history of alerts that come into the system.
You reimplemented this. In your approach you wait for alert in Firing state but you are not checking for Pending state.
Why didn't you reuse existing resources like measure_operation function in combination with fixtures as done in https://github.com/red-hat-storage/ocs-ci/blob/master/tests/functional/monitoring/conftest.py?

nagendra202 commented on ocs_ci/utility/prometheus.py:
```diff
@@ -783,3 +794,75 @@ def unsubscribe(self):
+    severity,
+):
+    """
+    Check all the properties like severity, state, message, description, runbook_url
```
made changes

nagendra202 commented on ocs_ci/utility/prometheus.py:
```diff
@@ -63,6 +63,17 @@ def check_alert_list(
             )
             assert len(found_alerts) == 1, assert_msg
 
+        assert_msg = f"Alert message for alert {label} is not correct"
```
made changes

nagendra202 commented on ocs_ci/templates/workloads/helper_scripts/meta_data_io.py:
```diff
@@ -0,0 +1,111 @@
+import time
```
done.

nagendra202 reviewed (commented):
addressed review comments.

nagendra202 reviewed (commented):
Replied to review comments.

nagendra202 commented:
@OdedViner  and @fbalak  I have addressed all review comments and stabilised code very well now. Please review and provide your feedback.  There were failures in PR job validation as part of teardown. I have taken care and rerunning the PR validation now.

nagendra202 reviewed (commented):
Addressed all review and stabilised code in tests and teardown. Please review it.
